pkg/controller/sparkapplication/controller.go

```diff
@@ -289,397 +300,472 @@ func (c *Controller) processNextItem() bool {
 	// There was a failure so be sure to report it. This method allows for pluggable error handling
 	// which can be used for things like cluster-monitoring
 	utilruntime.HandleError(fmt.Errorf("failed to sync SparkApplication %q: %v", key, err))
-	// Since we failed, we should requeue the item to work on later.  This method will add a backoff
-	// to avoid hot-looping on particular items (they're probably still not going to work right away)
-	// and overall controller protection (everything I've done is broken, this controller needs to
-	// calm down or it can starve other useful work) cases.
-	c.queue.AddRateLimited(key)
-
 	return true
 }
 
-func (c *Controller) syncSparkApplication(key string) error {
-	namespace, name, err := cache.SplitMetaNamespaceKey(key)
-	if err != nil {
-		return fmt.Errorf("failed to get the namespace and name from key %s: %v", key, err)
-	}
+// Helper Struct to encapsulate current State of the driver pod.
+type driverState struct {
+	podName            string         // Name of the driver pod.
+	sparkApplicationID string         // sparkApplicationID.
+	nodeName           string         // Name of the node the driver pod runs on.
+	podPhase           apiv1.PodPhase // Driver pod phase.
+	completionTime     metav1.Time    // Time the driver completes.
+}
 
-	app, err := c.getSparkApplication(namespace, name)
-	if err != nil {
-		return err
-	}
+func (c *Controller) getUpdatedAppStatus(app *v1alpha1.SparkApplication) *v1alpha1.SparkApplication {
+
+	// Fetch all the pods for the App.
+	selector, err := labels.NewRequirement(config.SparkAppNameLabel, selection.Equals, []string{app.Name})
```

Comment:
I feel that it's a bit insufficient to just use the app name as the label selector. To make sure that we just select pods of the current run, I think we should use a label that is unique for each run of an app and also add that into the label selector.